### PR TITLE
Statistics history task enhancement

### PR DESCRIPTION
**Pull request by Innovative-ashwin**
Signed-off-by: Innovative-ashwin <sinha.ashwini111@gmail.com>

### CODE REVIEW

**Review by gnandan — APPROVED** [review]
It's fine. Please test your changes for impact analysis before merging with develop

File: C/tasks/statistics_history/stats_history.cpp [hunk 1]
```diff
@@ -59,17 +59,30 @@ void StatsHistory::run() const
...
-			processKey(key);
+			processKey(key, historyValues, updateValues);
 		} catch (exception e) {
 			getLogger()->error("Failed to process statisitics key %s, %s", key, e.what());
 		}
                 rowIter = keySet->nextRow(rowIter);
 	} while (keySet->hasNextRow(rowIter));
 
+	int n_rows;
+        if ((n_rows = getStorageClient()->insertTable("statistics_history", historyValues)) != 1)
```
[the remark this answers is not in the document]
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp]: fixed

File: C/tasks/statistics_history/stats_history.cpp [hunk 2]
```diff
@@ -96,21 +109,14 @@ Query	query(new Where("key", Equals, key));
...
 	
 	// Insert the row into the configuration history
-	InsertValues historyValues;
 	historyValues.push_back(InsertValue("key", key.c_str()));
 	historyValues.push_back(InsertValue("value", val - prev));
 	historyValues.push_back(InsertValue("history_ts", "now()"));
-	int n_rows;
-	if ((n_rows = getStorageClient()->insertTable("statistics_history", historyValues)) != 1)
-	{
-		getLogger()->error("Failed to insert single row to statisitics history table for key %s", key.c_str());
```
[the remark this answers is not in the document]
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp:106]: fixed

File: C/tasks/statistics_history/stats_history.cpp [hunk 3]
```diff
@@ -96,21 +109,14 @@ Query	query(new Where("key", Equals, key));
...
-	}
+
 
 	// Update the previous value in the statistics row
-	InsertValues updateValues;
-	updateValues.push_back(InsertValue("previous_value", val));
-	if (getStorageClient()->updateTable("statistics", updateValues, Where("key", Equals, key)) != 1)
-	{
-		getLogger()->error("Failed to update single row to statisitics table for key %s", key.c_str());
```
[the remark this answers is not in the document]
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp:114]: fixed

File: scripts/tasks/statistics [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-# Run a Fledge south service written in Python
+# Run a Fledge south service written in C
```
Comment (praveen-garg) [inline scripts/tasks/statistics:2]: Not a Fledge south service but Fledge task

File: scripts/tasks/statistics [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-# Run a Fledge south service written in Python
+# Run a Fledge south task written in C
```
Comment (praveen-garg) [inline scripts/tasks/statistics:2]: south task?

**Comment by praveen-garg** [comment]
Can we avoid making commits for single line changes? There are 20+ commits in last 1 hour. Unless a miss we should keep commits atomic.

File: scripts/tasks/statistics (hunk 5 above)
Comment (Innovative-ashwin) [reply scripts/tasks/statistics]: fixed

**Review by MarkRiddoch — COMMENTED** [review]
Did you measure any performance improvements?

File: C/common/storage_client.cpp [hunk 6]
```diff
@@ -1742,3 +1742,162 @@ bool StorageClient::createSchema(const std::string& payload)
...
+                                                 it != values.cend(); ++it)
+                {
+                        if (it != values.cbegin())
+                        {
+                                convert << ", ";
+                        }
+                        convert <<  it->toJSON() ;
+                }
+
+                char url[128];
```
Comment (gnandan) [inline C/common/storage_client.cpp:1873]: is URL length 128 sufficient?

**Review by gnandan — COMMENTED** [review]
history_ts field should be  same for each set of keys BUFFERED,DISCARDED,PURGED,READINGS,SINUSOID,SN1-Ingest,UNSENT,UNSNPURGED

Please make sure you to generate history_ts correctly

File: C/tasks/statistics_history/stats_history.cpp [hunk 7]
```diff
@@ -94,23 +140,47 @@ Query	query(new Where("key", Equals, key));
...
-	}
+
+	InsertValue *updateValue = new InsertValue("previous_value", val);
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:157]: is not the variable name confusing?

Comment (praveen-garg) [reply C/tasks/statistics_history/stats_history.cpp:157]: We should leave inline comments, what is this doing.

File: C/tasks/statistics_history/stats_history.cpp [hunk 8]
```diff
@@ -59,33 +66,72 @@ void StatsHistory::run() const
...
+ * @param key	         The statistics key to process
+ * @param historyValues  Values to be inserted in statistics_history
+ * @param updateValues   Values to be updated in statistics
+ * @return void
  */
-void StatsHistory::processKey(const string& key) const
+void StatsHistory::processKey(const std::string& key, std::vector<InsertValues> &historyValues, std::vector<std::pair<InsertValue*, Where *> > &updateValues) const
 {
-Query	query(new Where("key", Equals, key));
+	Query	query(new Where("key", Equals, key));
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:128]: We should not query the statistics table in a loop for each key. Get all stats, find delta for each, put into stats history as a batch insert with same history_ts

File: C/tasks/statistics_history/stats_history.cpp [hunk 9]
```diff
@@ -54,63 +61,120 @@ void StatsHistory::run() const
 		return;
 
 	// Get the set of distinct statistics keys
-	Query query(new Returns("key"));
+	/*Query query(new Returns("key"));
 	query.distinct();
 	ResultSet *keySet = getStorageClient()->queryTable("statistics", query);
+	*/
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:67]: remove

File: C/tasks/statistics_history/stats_history.cpp [hunk 10]
```diff
@@ -54,63 +61,120 @@ void StatsHistory::run() const
...
-                rowIter = keySet->nextRow(rowIter);
-	} while (keySet->hasNextRow(rowIter));
+		if (!keySet->isLastRow(rowIter))
+                	rowIter = keySet->nextRow(rowIter);
+		else
+			break;
+	}
+
+	int n_rows;
+
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:100]: remove extra line

File: C/tasks/statistics_history/stats_history.cpp [hunk 11]
```diff
@@ -54,63 +61,120 @@ void StatsHistory::run() const
...
-	int n_rows;
-	if ((n_rows = getStorageClient()->insertTable("statistics_history", historyValues)) != 1)
-	{
-		getLogger()->error("Failed to insert single row to statisitics history table for key %s", key.c_str());
-	}
+	iValue.push_back(InsertValue("key", key.c_str()));
+	iValue.push_back(InsertValue("value", val - prev));
+	iValue.push_back(InsertValue("history_ts", dateTimeStr));
 
-	// Update the previous value in the statistics row
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:109]: keep this inline comment

File: C/tasks/statistics_history/stats_history.cpp [hunk 12]
```diff
@@ -10,7 +10,13 @@
 
 #include <stats_history.h>
 #include <csignal>
+#include <time.h>
+#include <sys/time.h>
 
+#define MAXSIZE 200
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:16]: what is this for?

File: C/tasks/statistics_history/stats_history.cpp [hunk 13]
```diff
@@ -54,63 +61,120 @@ void StatsHistory::run() const
...
+	std::vector<InsertValues> historyValues;
+	vector<pair<InsertValue *, Where *>> updateValues;
+
+	std::string dateTimeStr = getTime();
 
-        do {
+        while (keySet->hasNextRow(rowIter) || keySet->isLastRow(rowIter) )
+	{
 		string key = (*rowIter)->getColumn("key")->getString();
+		getLogger()->error("key = %s", key.c_str());
```
Comment (praveen-garg) [inline C/tasks/statistics_history/stats_history.cpp:84]: remove

File: C/common/storage_client.cpp (hunk 6 above)
Comment (Innovative-ashwin) [reply C/common/storage_client.cpp:1873]: fixed, made it 1000

File: C/tasks/statistics_history/stats_history.cpp (hunk 10 above)
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp:94]: fixed

File: C/tasks/statistics_history/stats_history.cpp (hunk 11 above)
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp:109]: fixed.

File: C/tasks/statistics_history/stats_history.cpp (hunk 12 above)
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp]: removed

**Comment by Innovative-ashwin** [comment]
> Did you measure any performance improvements?

earlier for a set of 10 keys , 30 calls were going to storage layer , now only 3

File: C/tasks/statistics_history/stats_history.cpp (hunk 8 above)
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp]: fixed

File: scripts/tasks/statistics (hunk 4 above)
Comment (Innovative-ashwin) [reply scripts/tasks/statistics]: fixed

File: C/tasks/statistics_history/stats_history.cpp [hunk 14]
```diff
@@ -56,61 +62,117 @@ void StatsHistory::run() const
...
-		getLogger()->error("Failed to update single row to statisitics table for key %s", key.c_str());
-	}
+	// create an object of InsertValue and push in updateValues vector
+	// for batch updation
+	InsertValue *updateValue = new InsertValue("previous_value", val);
+	Where *wKey = new Where("key", Equals, key);
+	updateValues.emplace_back(updateValue, wKey);
+}
+
+std::string StatsHistory::getTime(void) const
```
Comment (MarkRiddoch) [inline C/tasks/statistics_history/stats_history.cpp:154]: Should have a doxygen comment

**Review by MarkRiddoch — APPROVED** [review]

File: C/tasks/statistics_history/stats_history.cpp (hunk 14 above)
Comment (Innovative-ashwin) [reply C/tasks/statistics_history/stats_history.cpp:161]: fixed

**Comment by Innovative-ashwin** [comment]
Tested with 2 sinusoid south services  and OMF north service , with host config
Linux nerd-ThinkPad-L430 5.4.0-144-generic #161~18.04.1-Ubuntu SMP  x86_64 x86_64 x86_64 GNU/Linux

> In Python , on an average it takes 500 milliseconds for 1 iteration , however in C ( with new implementation) it takes 20 milliseconds to complete one iteration.

> with the python implementation , there are 3 calls to the storage layer client , one for querying statistics table , second to insert into statistics_history and third to update statistics.

With the previous implementation in C, the calls to storage layer was equal to 2 * number of keys ( one update and one insert per key)  plus one query to statistics, but with new implementation , there will be altogether 3 calls to the storage layer as in the python case.

**Pull request merged**